File: apps/faithful-ios/App/AppEnvironment.swift
import Foundation
import FaithfulKit

/// Where this build points, read once at launch from the bundle.
///
/// ## Why it fails closed
///
/// The origin comes from the `.xcconfig` for the configuration being built, and
/// **Staging and Release ship with it empty**. That is deliberate. A build whose
/// origin is missing does not fall back to production: it produces
/// `.unconfigured`, the app shows a state that says so, and the developer sees
/// exactly which key is absent.
///
/// The usual default is the other way round — a fallback to production — and it
/// is the reason staging builds end up writing to real churches. A build that
/// refuses to start is a bad afternoon; a staging build silently talking to
/// production is a bad year.
enum AppEnvironmentLoadResult {
    case configured(APIEnvironment, clientBuild: Int, allowsDebugControls: Bool)
    /// Which key was missing or unusable. Names the key, never a value.
    case unconfigured(reason: String)
}

enum AppEnvironmentLoader {
    /// Reads the environment from a bundle's Info dictionary.
    ///
    /// Takes the dictionary rather than reading `Bundle.main` directly so a test
    /// can drive every failure without building three apps.
    static func load(from info: [String: Any]) -> AppEnvironmentLoadResult {
        guard let key = string(info["FaithfulEnvironmentKey"]), !key.isEmpty else {
            return .unconfigured(reason: "FaithfulEnvironmentKey is not set")
        }

        guard let origin = string(info["FaithfulAPIOrigin"]), !origin.isEmpty else {
            return .unconfigured(reason: "FaithfulAPIOrigin is not set for \(key)")
        }

        guard let url = URL(string: origin), let scheme = url.scheme?.lowercased() else {
            return .unconfigured(reason: "FaithfulAPIOrigin is not a URL")
        }

        // A production or staging build talking over cleartext is a
        // configuration mistake, not a preference. Only a development build may.
        if scheme != "https" && key != "development" {
            return .unconfigured(reason: "FaithfulAPIOrigin must be https in \(key)")
        }
        if scheme != "https" && scheme != "http" {
            return .unconfigured(reason: "FaithfulAPIOrigin must be http or https")
        }

        // A build number that cannot be read is not defaulted: the server
        // refuses builds it no longer supports, and guessing one would make an
        // unsupported build look supported.
        guard let build = Int(string(info["FaithfulClientBuild"]) ?? ""), build > 0 else {
            return .unconfigured(reason: "FaithfulClientBuild is not a positive integer")
        }

        return .configured(
            APIEnvironment(key: key, baseURL: url),
            clientBuild: build,
            // Anything other than an explicit YES is NO. A misspelled value must
            // not switch developer affordances on in a build going to a church.
            allowsDebugControls: (string(info["FaithfulAllowDebugControls"]) ?? "NO") == "YES"
        )
    }

    private static func string(_ value: Any?) -> String? {
        guard let value = value as? String else { return nil }
        return value.trimmingCharacters(in: .whitespacesAndNewlines)
    }
}

File: apps/faithful-android/app/src/main/kotlin/io/faithform/faithful/AppEnvironment.kt
package io.faithform.faithful

/**
 * Where this build points, resolved once at launch.
 *
 * ## Why it fails closed
 *
 * The origin comes from the build type, and **staging and release ship with it
 * empty**. That is deliberate. A build whose origin is missing does not fall
 * back to production: it produces [Unconfigured], the app shows a state that
 * says so, and the developer sees exactly which value is absent.
 *
 * The usual default is the other way round — a fallback to production — and it
 * is the reason staging builds end up writing to real churches. This file used
 * to be that: `release` hardcoded `https://faithform.io`.
 *
 * Mirrors `AppEnvironment.swift` on iOS, decision for decision.
 */
sealed interface AppEnvironment {
    data class Configured(
        val environmentKey: String,
        val apiOrigin: String,
        val clientBuild: Int,
        val allowsDebugControls: Boolean,
    ) : AppEnvironment

    /** Names the value that is missing. Never a value itself. */
    data class Unconfigured(val reason: String) : AppEnvironment
}

object AppEnvironmentLoader {

    fun load(
        environmentKey: String,
        apiOrigin: String,
        clientBuild: Int,
        allowDebugControls: Boolean,
    ): AppEnvironment {
        if (environmentKey.isBlank()) {
            return AppEnvironment.Unconfigured("ENVIRONMENT_KEY is not set")
        }
        val origin = apiOrigin.trim()
        if (origin.isEmpty()) {
            return AppEnvironment.Unconfigured("API_ORIGIN is not set for $environmentKey")
        }

        val scheme = origin.substringBefore("://", missingDelimiterValue = "").lowercase()
        if (scheme != "https" && scheme != "http") {
            return AppEnvironment.Unconfigured("API_ORIGIN must be http or https")
        }
        // A production or staging build talking over cleartext is a
        // configuration mistake, not a preference. Only a development build may
        // — and the manifest forbids cleartext outside the emulator loopback
        // anyway, so this is the second of two independent refusals.
        if (scheme != "https" && environmentKey != "development") {
            return AppEnvironment.Unconfigured("API_ORIGIN must be https in $environmentKey")
        }

        // A build number that cannot be read is not defaulted: the server
        // refuses builds it no longer supports, and guessing one would make an
        // unsupported build look supported.
        if (clientBuild <= 0) {
            return AppEnvironment.Unconfigured("VERSION_CODE is not a positive integer")
        }

        return AppEnvironment.Configured(
            environmentKey = environmentKey,
            apiOrigin = origin,
            clientBuild = clientBuild,
            allowsDebugControls = allowDebugControls,
        )
    }
}
